### DISCUSSION TITLE
Why TrainState can be use as arguments of jited function without trigger error?

### DISCUSSION THREAD

**Discussion by wztdream** [google/flax#1858] [General] [answered]
Hi,
Just for curiosity, as [jit doc](https://github.com/google/jax/blob/0e92124c5b0c9b2ab4ed306dc4491413062fd3db/jax/api.py#L120) says the arguments of a jit-ed function must be:

> Its arguments and return value should be arrays,
>  scalars, or (nested) standard Python containers (tuple/list/dict) thereof

But TrainState can be use in jit-ed function without trigger any error, for example:
```
@functools.partial(jax.jit, static_argnums=(2, 3))
def train_step(state, batch, loss_fn, train_metrics_fn, rngs):
    # some train step code here,state is an instance of TrainState
```
We have to mark `loss_fn` which is a Callable as static arguments to avoid error, but no need for`state` which is also violate the constrains in jit doc, so why there is no error for state?

**Comment by marcvanzee** [comment, marked as answer]
When you look at the [definition](https://github.com/google/flax/blob/main/flax/training/train_state.py#L22) of the class `TrainState`, you can see that it extends `struct.PyTreeNode`. This means that it is a dataclasses that acts act like a JAX pytree node, so JAX knows how to flatten/unflatten it for JAX transformations.
